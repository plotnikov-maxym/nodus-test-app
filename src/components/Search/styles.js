import {fade} from "@material-ui/core/styles/colorManipulator";

export const styles = ({palette, spacing: {unit}}) => ({
  root: {
    padding: unit * 3,
  },
  grow: {
    flexGrow: 1,
  },
  container: {
    display: "flex",
    alignItems: "center",
  },
  inputContainer: {
    minWidth: "50%",
    marginRight: unit * 2,
  },
  search: {
    backgroundColor: fade(palette.common.white, 0.15),
    "&:hover": {
      backgroundColor: fade(palette.common.white, 0.25),
    },
  },
  button: {
    marginLeft: unit * 2,
  },
});
