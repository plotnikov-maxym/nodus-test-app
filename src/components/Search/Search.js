import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Formik, Field, Form} from 'formik';
import {TextField} from 'formik-material-ui';
import FormControl from '@material-ui/core/FormControl';
import SearchIcon from '@material-ui/icons/Search';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import {withStyles} from '@material-ui/core/styles';
import {locale} from '../../constants/locales';
import {styles} from './styles';

class Search extends Component {
  state = {
    value: '',
  };

  render() {
    const {classes} = this.props;
    return (
        <Formik
            initialValues={{
              search: '',
            }}
            onSubmit={(values, {setSubmitting}) => {
              setTimeout(() => {
                setSubmitting(false);
                alert(JSON.stringify(values, null, 2));
              }, 1000);
            }}
            render={({submitForm, isSubmitting, values, setFieldValue}) => (
                <Form className={classes.root}>
                  <div className={classes.container}>
                    <FormControl>
                      <Grid container spacing={8} alignItems="flex-end">
                        <Grid item>
                          <SearchIcon/>
                        </Grid>
                        <Grid item>
                          <div className={classes.search}>
                            <Field
                                type="text"
                                label={locale.SEARCH_IN_PUBMED}
                                placeholder={locale.SEARCH}
                                name="search"
                                component={TextField}
                                margin="none"
                                fullWidth
                            />
                          </div>
                        </Grid>
                      </Grid>
                    </FormControl>

                    <Button type="submit" variant="contained" color="primary"
                            className={classes.button}>
                      {locale.SEARCH}
                    </Button>

                    <Button variant="contained" color="primary"
                            className={classes.button}>
                      {locale.ADD_TO_FAVOURITES}
                    </Button>
                  </div>
                </Form>
            )}
        />

    );
  };
}

Search.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(Search);
