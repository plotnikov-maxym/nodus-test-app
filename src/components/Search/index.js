import {connect} from "react-redux";
import {fetchArticles} from "../../actions/articles";
import {addToFavourites} from "../../actions/favourites";
import Search from "./Search";

const mapDispatchToProps = dispatch => ({
  fetchArticles: query => dispatch(fetchArticles(query)),
  addToFavourites: query => dispatch(addToFavourites(query)),
});
export default connect(
  () => ({}),
  mapDispatchToProps,
)(Search);
