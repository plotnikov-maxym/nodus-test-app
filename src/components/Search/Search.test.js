import React from "react";
import renderer from "react-test-renderer";
import {extend} from "underscore";
import {Search} from "./Search";
import {createShallow} from "@material-ui/core/test-utils";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";

describe("<Search />", () => {
  let shallow;

  beforeAll(() => {
    shallow = createShallow();
  });

  describe("Search component", () => {
    it("matches snapshot", async () => {
      const wrapper = await renderComponent();
      expect(wrapper).toMatchSnapshot();
    });

    it("renders the correctly Input, changes value, and renders Delete button", async () => {
      const searchQuery = "Test Search Query";
      const wrapper = await getComponent();
      const input = wrapper.find(TextField);
      const inputAdornmentBefore = wrapper.find(TextField).prop("InputProps")
        .endAdornment;
      const fieldOnChange = input.prop("onChange");
      fieldOnChange({
        target: {name: "search", value: searchQuery},
      });
      wrapper.update();
      expect(wrapper.find(TextField).prop("value")).toBe(searchQuery);
      const inputAdornmentAfter = wrapper.find(TextField).prop("InputProps")
        .endAdornment;
      expect(inputAdornmentBefore).not.toBe(inputAdornmentAfter);
      const btn = shallow(inputAdornmentAfter);
      const icon = shallow(btn.props().children);
      icon.prop("onClick")();
      wrapper.update();
      expect(wrapper.find(TextField).prop("value")).toBe("");
    });

    it("renders submit button and handles form submission with fetchArticles", async () => {
      const fetchArticles = jest.fn();
      const searchQuery = "Test Search Query";
      const wrapper = await getComponent({fetchArticles});
      const input = wrapper.find(TextField);
      const fieldOnChange = input.prop("onChange");
      fieldOnChange({
        target: {name: "search", value: searchQuery},
      });
      const submitButton = wrapper
        .find(Button)
        .findWhere(btn => btn.prop("type") === "submit");
      expect(submitButton).toExist();
      const form = wrapper.find("form");
      form.prop("onSubmit")({
        preventDefault: () => {},
      });
      expect(wrapper.state("value")).toBe(searchQuery);
      await expect(fetchArticles).toHaveBeenCalledWith(searchQuery);
    });
    it("does not adding to favourites if value is less than on character", async () => {
      const addToFavourites = jest.fn();
      const searchQuery = "a";
      const wrapper = await getComponent({addToFavourites});
      const input = wrapper.find(TextField);
      const fieldOnChange = input.prop("onChange");
      fieldOnChange({
        target: {name: "search", value: searchQuery},
      });
      wrapper.update();
      const addButton = wrapper
        .find(Button)
        .findWhere(btn => btn.prop("type") === "button");
      addButton.simulate("click");
      await expect(addToFavourites).not.toHaveBeenCalled();
    });

    it("renders add to favourites button and handles adding to favourites function", async () => {
      const addToFavourites = jest.fn();
      const searchQuery = "Test Query";
      const expectedQuery = {id: "Test_Query", value: searchQuery};
      const wrapper = await getComponent({addToFavourites});
      const input = wrapper.find(TextField);
      const fieldOnChange = input.prop("onChange");
      fieldOnChange({
        target: {name: "search", value: searchQuery},
      });
      wrapper.update();
      const addButton = wrapper
        .find(Button)
        .findWhere(btn => btn.prop("type") === "button");
      expect(addButton).toExist();
      addButton.simulate("click");
      await expect(addToFavourites).toHaveBeenCalledWith(expectedQuery);
    });
  });

  const getComponent = props => {
    const parsedProps = extend(getMockProps(), props);

    return shallow(<Search {...parsedProps} />);
  };

  const renderComponent = props => {
    const parsedProps = extend(getMockProps(), props);

    return renderer.create(<Search {...parsedProps} />).toJSON();
  };

  const getMockProps = () => ({
    classes: {
      button: "Search-button-84",
      container: "Search-container-81",
      grow: "Search-grow-80",
      inputContainer: "Search-inputContainer-82",
      root: "Search-root-79",
      search: "Search-search-83",
    },
    fetchArticles: () => {},
    addToFavourites: () => {},
  });
});
