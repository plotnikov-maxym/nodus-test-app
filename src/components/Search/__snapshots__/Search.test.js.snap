// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> Search component matches snapshot 1`] = `
<form
  className="Search-root-79"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="Search-container-81"
  >
    <div
      className="MuiFormControl-root-1 Search-inputContainer-82"
    >
      <div
        className="MuiFormControl-root-1 Search-search-83"
      >
        <label
          className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12 MuiInputLabel-outlined-15"
          data-shrink={true}
        >
          Search in PubMed
        </label>
        <div
          className="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37 MuiInputBase-adornedStart-40 MuiOutlinedInput-adornedStart-26 MuiInputBase-adornedEnd-41 MuiOutlinedInput-adornedEnd-27"
          onClick={[Function]}
        >
          <fieldset
            aria-hidden={true}
            className="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
            style={
              Object {
                "paddingLeft": 8,
              }
            }
          >
            <legend
              className="MuiPrivateNotchedOutline-legend-54"
              style={
                Object {
                  "width": 0,
                }
              }
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </legend>
          </fieldset>
          <div
            className="MuiInputAdornment-root-55 MuiInputAdornment-positionStart-57"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-60"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
          <input
            aria-invalid={false}
            className="MuiInputBase-input-46 MuiOutlinedInput-input-31 MuiInputBase-inputAdornedStart-51 MuiOutlinedInput-inputAdornedStart-34 MuiInputBase-inputAdornedEnd-52 MuiOutlinedInput-inputAdornedEnd-35"
            disabled={false}
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search"
            required={false}
            type="text"
            value=""
          />
          <div
            className="MuiInputAdornment-root-55 MuiInputAdornment-positionEnd-58"
          />
        </div>
      </div>
    </div>
    <button
      className="MuiButtonBase-root-95 MuiButton-root-69 MuiButton-contained-80 MuiButton-containedPrimary-81 MuiButton-raised-83 MuiButton-raisedPrimary-84 Search-button-84"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-70"
      >
        Search
      </span>
      <span
        className="MuiTouchRipple-root-98"
      />
    </button>
    <button
      className="MuiButtonBase-root-95 MuiButton-root-69 MuiButton-contained-80 MuiButton-containedPrimary-81 MuiButton-raised-83 MuiButton-raisedPrimary-84 Search-button-84"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-70"
      >
        Add to favourites
      </span>
      <span
        className="MuiTouchRipple-root-98"
      />
    </button>
  </div>
</form>
`;
